File: src/main/java/destiny/utils/Base58.java
/**
 * @author smallufo Created on 2011/3/11 at 上午4:08:05
 */
package destiny.utils;

import java.io.Serializable;

/**
 * http://java.dzone.com/articles/create-your-own-bitly-using
 */
public class Base58 implements Serializable
{
  private static final char[] BASE58_CHARS = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz".toCharArray();

  public static String numberToAlpha(long number)
  {
    char[] buffer = new char[20];
    int index = 0;
    do
    {
      buffer[index++] = BASE58_CHARS[(int) (number % BASE58_CHARS.length)];
      number = number / BASE58_CHARS.length;
    }
    while (number > 0);
    return new String(buffer, 0, index);
  }

  public static long alphaToNumber(String text)
  {
    char[] chars = text.toCharArray();
    long result = 0;
    long multiplier = 1;
    for (int index = 0; index < chars.length; index++)
    {
      char c = chars[index];
      int digit;
      if (c >= '1' && c <= '9')
      {
        digit = c - '1';
      }
      else if (c >= 'A' && c < 'I')
      {
        digit = (c - 'A') + 9;
      }
      else if (c > 'I' && c < 'O')
      {
        digit = (c - 'J') + 17;
      }
      else if (c > 'O' && c <= 'Z')
      {
        digit = (c - 'P') + 22;
      }
      else if (c >= 'a' && c < 'l')
      {
        digit = (c - 'a') + 33;
      }
      else if (c > 'l' && c <= 'z')
      {
        digit = (c - 'l') + 43;
      }
      else
      {
        throw new IllegalArgumentException("Illegal character found: '" + c + "'");
      }

      result += digit * multiplier;
      multiplier = multiplier * BASE58_CHARS.length;
    }
    return result;
  }

}

File: src/main/java/destiny/tools/Base58.kt
/**
 * @author smallufo Created on 2011/3/11 at 上午4:08:05
 */
package destiny.tools

import java.io.Serializable

/**
 * http://java.dzone.com/articles/create-your-own-bitly-using
 */
class Base58 : Serializable {
  companion object {

    private val BASE58_CHARS = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz".toCharArray()

    fun numberToAlpha(number: Long): String {
      var number = number
      val buffer = CharArray(20)
      var index = 0
      do {
        buffer[index++] = BASE58_CHARS[(number % BASE58_CHARS.size).toInt()]
        number /= BASE58_CHARS.size
      } while (number > 0)
      return String(buffer, 0, index)
    }

    fun alphaToNumber(text: String): Long {
      val chars = text.toCharArray()
      var result: Long = 0
      var multiplier: Long = 1
      for (c in chars) {
        val digit: Int = when (c) {
          in '1'..'9' -> c - '1'
          in 'A'..'H' -> c - 'A' + 9
          in 'J'..'N' -> c - 'J' + 17
          in 'P'..'Z' -> c - 'P' + 22
          in 'a'..'k' -> c - 'a' + 33
          in 'm'..'z' -> c - 'l' + 43
          else -> throw IllegalArgumentException("Illegal character found: '$c'")
        }

        result += digit * multiplier
        multiplier *= BASE58_CHARS.size
      }
      return result
    }
  }

}
